src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncContext.Unix.cs

```diff
@@ -2023,15 +2029,35 @@ public Interop.Sys.SocketEvents HandleSyncEventsSpeculatively(Interop.Sys.Socket
             return events;
         }
 
-        public unsafe void HandleEvents(Interop.Sys.SocketEvents events)
+        // Called on the epoll thread.
+        public Interop.Sys.SocketEvents HandleSyncAndInlineEvents(Interop.Sys.SocketEvents events)
         {
             if ((events & Interop.Sys.SocketEvents.Error) != 0)
             {
-                // Set the Read and Write flags as well; the processing for these events
+                // Set the Read and Write flags; the processing for these events
                 // will pick up the error.
+                events ^= Interop.Sys.SocketEvents.Error;
                 events |= Interop.Sys.SocketEvents.Read | Interop.Sys.SocketEvents.Write;
             }
 
+            if ((events & Interop.Sys.SocketEvents.Read) != 0 &&
+                _receiveQueue.ProcessSyncEventOrGetAsyncEvent(this, inlineAsyncEvents: true) == null)
```

Comment:
This is now hard coded to inline. If configurable per Socket/AsyncOperation we'd read a property.

Reply:
Maybe to enable experimentation at the ASP.NET layer we should add the Socket property now, but as internal (e.g. `internal bool PreferInlineCompletions` or something like that).  We can use reflection to set it from ASP.NET and then decide whether to make it public or rip it out. We'll just need to make sure that trimming doesn't remove it, so we'd need to add a [PreserveDependency(...)] attribute from something that won't be trimmed to it.  Adding that now will also make it easier for us to remember how to do it and what to edit later on :)

If you'd prefer to do it separately, that's fine, too.

Reply:
I'll add the `PreferInlineCompletions` property in this PR.